bpf/bpf_host.c

```diff
@@ -997,20 +997,31 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
 	}
 #endif
 
+	/**
+	 * Trace packets destined to the wire here, since our SNAT implementation
+	 * involves a tail call and this is the last common code path. This should be
+	 * improved when tail calls can be removed in favor of BPF-to-BPF.
+	 */
+	send_trace_notify(ctx, TRACE_TO_NETWORK, src_id, 0, 0,
+			  0, ret, monitor);
+
```

Comment:
I don't think it's a great idea to move this before the NAT, because then the trace notification that is emitted will not represent the final state of the packet when the packet leaves Cilium context. For instance, we will not know how NAT was applied. (Admittedly at the moment we don't trace such messages at all, but still, I don't think this is quite getting to where we want to be around notifications).

Realistically to solve monitor aggregation for the NAT case as well, I think we need to pass the `monitor` value through to the tail call (via `skb->cb[]` (preferred) or `skb->mark`) and then ensure that before that program returns the TC result, we perform tracing using the monitor value.

Reply:
Okay, I wasn't sure how to implement this to begin with, but this would indeed be a great feature to have from the user's perspective. I'll take a stab at it. :+1:

Reply:
I've added a naive implementation of this in 0e4192a1af281327f4383332407a18209203c9a4. I have some general questions I'll add as a review to this PR.